### PR TITLE
Enable Selecting micro:bit on Levelbuilder levels

### PR DESCRIPTION
**Pull request by epeach**
Reverts code-dot-org/code-dot-org#48234

Re-applying this change, while taking in to account this failure that caused the revert: https://codedotorg.slack.com/archives/C0T0PNTM3/p1663855722451139

This had to be reverted because it wasn't properly handling the old property type in levels where `makerlab_enabled` was true/false. In these cases, the values weren't accurately displayed in the dropdown. This was then creating an error because the value of `makerlab_enabled` was not matching the palette category. This PR also updates the error message to take into account that there are now three `makerlab_enabled` options (' ', 'circuitPlayground', and 'microbit') and validates accordingly.

Screenshot of updated error message: 
Screenshot from 2022-09-22 15-38-08

### CODE REVIEW

File: dashboard/app/models/levels/applab.rb [hunk 1]
```diff
@@ -125,12 +129,15 @@ def update_json_fields
   end
 
   def validate_maker_if_needed
-    maker_enabled = properties['makerlab_enabled'] == 'true'
+    makerlab_property = properties['makerlab_enabled']
```
Comment (bencodeorg) [inline dashboard/app/models/levels/applab.rb]: Out of scope for this PR, but would it be difficult backwards compatibility-wise to change the name of this property to something that indicates it's not a boolean? Tricky to read with the current naming

File: dashboard/app/models/levels/applab.rb [hunk 2]
```diff
@@ -74,7 +74,11 @@ def self.palette_categories
   end
 
   def self.maker_palette_categories
-    %w(Maker Circuit)
+    %w(Maker Circuit micro:bit)
```
Comment (bencodeorg) [inline dashboard/app/models/levels/applab.rb]: Could you add a comment describing the distinction between palette categories and APIs?

Comment (bencodeorg) [reply dashboard/app/models/levels/applab.rb]: Tangentially, is there a mapping between palette categories and APIs, or are they completely distinct? Wondering if there's a way to simplify the error handling logic that looks like this:

```
(starting_category == 'Circuit' && makerlab_property != 'circuitPlayground') ||
      (starting_category == 'micro:bit' && makerlab_property != 'microbit') ||
      (starting_category == 'Maker' && !maker_enabled)
```

File: dashboard/app/models/levels/applab.rb (hunk 1 above)
Comment (epeach) [reply dashboard/app/models/levels/applab.rb]: This is a good call. Elsewhere we use makerAPIEnabled (as in 'which' makerAPIEnabled). Does that sound like enough in the right direction?

File: dashboard/app/models/levels/applab.rb (hunk 2 above)
Comment (epeach) [reply dashboard/app/models/levels/applab.rb]: Oof, I definitely was wondering about this. The starting category is as displayed to the user and the makerlab_property is as saved to the main.json, so a bit different user case. I don't think we can align them perfectly, but we could probably make a look-up between the two to minimize potential error here.

File: dashboard/app/models/levels/applab.rb (hunk 1 above)
Comment (bencodeorg) [reply dashboard/app/models/levels/applab.rb]: huh, I find that confusing also (ie, why not just `makerAPI`?). Could be a just me thing. In any case, I think fine for now, but maybe something to consider as we move to multiple APIs?

**Comment by epeach** [comment]
@bencodeorg - PTAL!

File: dashboard/app/models/levels/applab.rb [hunk 3]
```diff
@@ -125,12 +129,25 @@ def update_json_fields
   end
 
   def validate_maker_if_needed
-    maker_enabled = properties['makerlab_enabled'] == 'true'
+    maker_api_property = properties['makerlab_enabled']
+    maker_enabled = SharedConstants::MAKER_PALETTE_CATEGORIES.keys.include?(maker_api_property)
```
Comment (bencodeorg) [inline dashboard/app/models/levels/applab.rb]: Maybe one thing I am finding confusing about this validation is that it doesn't seem to explicitly handle the old true/false format?

ie, if we made any edit to a legacy maker enabled level (outside of a UI), and ran this validation while saving it, the variable `maker_enabled` would be false (even if `properties['makerlab_enabled']` is set to true)? Then we'd raise an error below? I think this is handled in the HAML change you have here, but would we want to be able to handle non-UI edits to levels?

File: dashboard/app/models/levels/applab.rb [hunk 4]
```diff
@@ -125,12 +129,25 @@ def update_json_fields
...
     starting_category = properties['palette_category_at_start']
-    if Applab.maker_palette_categories.include?(starting_category) && !maker_enabled
+
+    # If the "Maker" palette category is selected, either Maker API must be enabled.
+    if starting_category == 'Maker' && !maker_enabled
```
Comment (bencodeorg) [inline dashboard/app/models/levels/applab.rb]: I think this is easier for me to reason about if the first check handles the cases where a level does not have maker enabled, and has any of the three hardware-related starting categories (not just Maker). And the end of the error message could be "...but this level does not have a Maker API enabled".

Then, the second check handles the specifics of picking a circuit or microbit starting category (which you're already doing).

File: dashboard/app/models/levels/applab.rb [hunk 5]
```diff
@@ -125,12 +129,25 @@ def update_json_fields
...
+      raise ArgumentError.new(
+        "Selected '#{starting_category}' as the palette category at start, " \
+            "but this level does not have either the circuitPlayground or micro:bit Maker API enabled."
+      )
+    end
+
+    # If a board-specific maker palette ('Circuit' or 'micro:bit') is selected and the corresponding
+    # maker API must also be selected
+    if ['Circuit', 'micro:bit'].include?(starting_category) &&
+      maker_api_property != SharedConstants::MAKER_PALETTE_CATEGORIES.key(starting_category)
```
Comment (bencodeorg) [inline dashboard/app/models/levels/applab.rb]: Seems a little odd to get the key from the value, maybe the mapping is backwards here? Looks like this is the only place we do a lookup, so maybe reverse? 

Tangentially, if I understand correctly, it looks like the key in `'maker' => "Maker"` isn't ever used, I wonder if we should remove from the hash and add add "Maker" in `self.maker_palette_categories`

**Review by bencodeorg — COMMENTED** [review]
Couple more questions for you, sorry to be a bother! Almost there I think!

File: dashboard/app/models/levels/applab.rb (hunk 3 above)
Comment (epeach) [reply dashboard/app/models/levels/applab.rb]: Yeah, with the HAML change, we won't hit this problem through the levelbuilder editing UI. (line 17 on '_droplet.html.haml' in this PR)

When an old API level is loaded and saved, the property looks like this: 
1. Before levelbuilder edit page is loaded: `properties['makerlab_enabled'] == true`.
2. When levelbuilder edit page is loaded and the above haml is rendered: `properties['makerlab_enabled'] == 'circuitPlayground'`.
3. User clicks 'save level' button.
4. Validation is run as part of saving and `properties['makerlab_enabled'] == 'circuitPlayground'` is the value used to validate/save.

I'm not following your question about 'outside of a UI' and 'non-UI edits to levels'? Can you describe that process?

File: dashboard/app/models/levels/applab.rb (hunk 5 above)
Comment (epeach) [reply dashboard/app/models/levels/applab.rb]: This was just a really messy way for me to consolidate all the palette categories into one place so I could reuse it in `self.maker_palette_categories` and minimize the number of times we have strings representing these values. In this case, I grabbed all the values of this to populate the Palette Categories.

However, I realized that the other palette categories were ids and not 'as-they-appear-to-students', so I did some digging and realized that I can do the same - use ids here and have 'friendly' names displayed elsewhere. I went ahead and made those changes, which also beautified the boolean. Hopefully, that all looks good now, but let me know if these concerns remain.

File: dashboard/app/models/levels/applab.rb (hunk 4 above)
Comment (epeach) [reply dashboard/app/models/levels/applab.rb]: Alright, I went ahead and switched around that first check. Let me know if that reads better.

Comment (bencodeorg) [reply dashboard/app/models/levels/applab.rb]: Looks great, thanks!

File: dashboard/app/models/levels/applab.rb (hunk 5 above)
Comment (bencodeorg) [reply dashboard/app/models/levels/applab.rb]: Sounds like a great simplification! Sorry for suggesting the mapping idea in the first place, probably was overkill.

File: dashboard/app/models/levels/applab.rb (hunk 3 above)
Comment (bencodeorg) [reply dashboard/app/models/levels/applab.rb]: Oh I just meant if we ever needed to bulk edit levels via a script or something, we wouldn't be able to save edits to these legacy-style levels.

Maybe more importantly, there's now no documentation on the model about the fact that we have lots of old levels with true/false for the makerlab_enabled property (ie, the documentation to this effect is in the HAML file you updated). I think we could resolve this by a) adding a comment in the model validation similar to what you have in the HAML file, and b) update the `maker_enabled` variable used in validation to something like this:

`maker_enabled = Applab.maker_apis.include?(maker_api_property) || maker_api_property == 'true'`

Comment (epeach) [reply dashboard/app/models/levels/applab.rb]: Ahh ha! I wasn't considering updating by script - thank you for reminding me that! Added the suggested changes.

**Review by bencodeorg — APPROVED** [review]
LGTM -- thanks for graciously tolerating my many questions!

**Pull request merged**